File: vespalib/src/vespa/vespalib/net/tls/capability_set.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "capability_set.h"
#include <vespa/vespalib/stllike/hash_map.hpp>
#include <vespa/vespalib/stllike/asciistream.h>
#include <cassert>

namespace vespalib::net::tls {

string CapabilitySet::to_string() const {
    asciistream os;
    os << "CapabilitySet({";
    bool emit_comma = false;
    for_each_capability([&emit_comma, &os](Capability cap) {
       if (emit_comma) {
           os << ", ";
       } else {
           emit_comma = true;
       }
       // TODO let asciistream and std::string_view play along
       os << stringref(cap.name().data(), cap.name().size());
    });
    os << "})";
    return os.str();
}

std::optional<CapabilitySet> CapabilitySet::find_capability_set(const string& cap_set_name) noexcept {
    static const hash_map<string, CapabilitySet> name_to_cap_set({
        {"vespa.all",                     all()},
        {"vespa.content_node",            content_node()},
        {"vespa.container_node",          container_node()},
        {"vespa.telemetry",               telemetry()},
        {"vespa.cluster_controller_node", cluster_controller_node()},
        {"vespa.logserver_node",          logserver_node()},
        {"vespa.config_server",           config_server()}
    });
    auto iter = name_to_cap_set.find(cap_set_name);
    return (iter != name_to_cap_set.end()) ? std::optional<CapabilitySet>(iter->second) : std::nullopt;
}

bool CapabilitySet::resolve_and_add(const string& set_or_cap_name) noexcept {
    if (auto cap_set = find_capability_set(set_or_cap_name)) {
        _capability_mask |= cap_set->_capability_mask;
        return true;
    } else if (auto cap = Capability::find_capability(set_or_cap_name)) {
        _capability_mask |= cap_as_bit_set(*cap);
        return true;
    }
    return false;
}

// Note: the capability set factory functions below are all just using constexpr and/or inline
// functions, so the compiler will happily optimize them to just "return <constant bit pattern>".

CapabilitySet CapabilitySet::all() noexcept {
    return make_with_all_capabilities();
}

CapabilitySet CapabilitySet::content_node() noexcept {
    return CapabilitySet::of({Capability::content_storage_api(),
                              Capability::content_document_api(),
                              Capability::container_document_api()})
            .union_of(shared_app_node_capabilities());
}

CapabilitySet CapabilitySet::container_node() noexcept {
    return CapabilitySet::of({Capability::content_document_api(),
                              Capability::container_document_api(),
                              Capability::content_search_api()})
            .union_of(shared_app_node_capabilities());
}

CapabilitySet CapabilitySet::telemetry() noexcept {
    return CapabilitySet::of({Capability::content_status_pages(),
                              Capability::content_metrics_api(),
                              Capability::container_state_api(),
                              Capability::metricsproxy_metrics_api(),
                              Capability::sentinel_connectivity_check()});
}

CapabilitySet CapabilitySet::cluster_controller_node() noexcept {
    return CapabilitySet::of({Capability::content_cluster_controller_internal_state_api(),
                              Capability::client_slobrok_api(),
                              Capability::container_document_api()}) // for reindexing
            .union_of(shared_app_node_capabilities());
}

CapabilitySet CapabilitySet::logserver_node() noexcept {
    return shared_app_node_capabilities();
}

CapabilitySet CapabilitySet::config_server() noexcept {
    return CapabilitySet::of({Capability::client_filereceiver_api(),
                              Capability::container_management_api(),
                              Capability::slobrok_api(),
                              Capability::cluster_controller_reindexing(),
                              Capability::cluster_controller_state(),
                              Capability::logserver_api()})
            .union_of(telemetry());
}

CapabilitySet CapabilitySet::shared_app_node_capabilities() noexcept {
    return CapabilitySet::of({Capability::logserver_api(),
                              Capability::configserver_config_api(),
                              Capability::configserver_filedistribution_api(),
                              Capability::configproxy_config_api(),
                              Capability::configproxy_filedistribution_api(),
                              Capability::slobrok_api()})
            .union_of(telemetry());
}

CapabilitySet CapabilitySet::make_with_all_capabilities() noexcept {
    BitSet bit_set;
    bit_set.flip(); // All cap bits set
    return CapabilitySet(bit_set);
}

std::ostream& operator<<(std::ostream& os, const CapabilitySet& cap_set) {
    os << cap_set.to_string();
    return os;
}

asciistream& operator<<(asciistream& os, const CapabilitySet& cap_set) {
    os << cap_set.to_string();
    return os;
}

}

File: security-utils/src/main/java/com/yahoo/security/tls/CapabilitySet.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.security.tls;

import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.SortedSet;
import java.util.TreeSet;
import java.util.logging.Logger;
import java.util.stream.Collectors;

/**
 * @author bjorncs
 */
public class CapabilitySet implements ToCapabilitySet {

    private static final Logger log = Logger.getLogger(CapabilitySet.class.getName());

    private static final Map<String, CapabilitySet> PREDEFINED = new HashMap<>();


    /* Predefined capability sets */
    public static final CapabilitySet ALL = predefined(
            "vespa.all", Capability.values());
    public static final CapabilitySet TELEMETRY = predefined(
            "vespa.telemetry",
            Capability.CONTENT__STATUS_PAGES, Capability.CONTENT__METRICS_API, Capability.CONTAINER__STATE_API,
            Capability.METRICSPROXY__METRICS_API, Capability.SENTINEL__CONNECTIVITY_CHECK);

    private static final CapabilitySet SHARED_CAPABILITIES_APP_NODE = CapabilitySet.of(
            Capability.LOGSERVER_API, Capability.CONFIGSERVER__CONFIG_API,
            Capability.CONFIGSERVER__FILEDISTRIBUTION_API, Capability.CONFIGPROXY__CONFIG_API,
            Capability.CONFIGPROXY__FILEDISTRIBUTION_API, Capability.SLOBROK__API, TELEMETRY);

    public static final CapabilitySet CONTENT_NODE = predefined(
            "vespa.content_node",
            Capability.CONTENT__STORAGE_API, Capability.CONTENT__DOCUMENT_API, Capability.CONTAINER__DOCUMENT_API,
            SHARED_CAPABILITIES_APP_NODE);
    public static final CapabilitySet CONTAINER_NODE = predefined(
            "vespa.container_node",
            Capability.CONTAINER__DOCUMENT_API, Capability.CONTENT__DOCUMENT_API, Capability.CONTENT__SEARCH_API,
            SHARED_CAPABILITIES_APP_NODE);
    public static final CapabilitySet CLUSTER_CONTROLLER_NODE = predefined(
            "vespa.cluster_controller_node",
            Capability.CONTENT__CLUSTER_CONTROLLER__INTERNAL_STATE_API,
            Capability.CLIENT__SLOBROK_API, Capability.CONTAINER__DOCUMENT_API, SHARED_CAPABILITIES_APP_NODE);
    public static final CapabilitySet LOGSERVER_NODE = predefined(
            "vespa.logserver_node", SHARED_CAPABILITIES_APP_NODE);
    public static final CapabilitySet CONFIGSERVER_NODE = predefined(
            "vespa.config_server_node",
            Capability.CLIENT__FILERECEIVER_API, Capability.CONTAINER__MANAGEMENT_API, Capability.SLOBROK__API,
            Capability.CLUSTER_CONTROLLER__REINDEXING, Capability.CLUSTER_CONTROLLER__STATE, Capability.LOGSERVER_API,
            TELEMETRY);

    private static CapabilitySet predefined(String name, ToCapabilitySet... capabilities) {
        var instance = CapabilitySet.of(capabilities);
        PREDEFINED.put(name, instance);
        return instance;
    }

    private static final CapabilitySet ALL_CAPABILITIES = new CapabilitySet(EnumSet.allOf(Capability.class));
    private static final CapabilitySet NO_CAPABILITIES = new CapabilitySet(EnumSet.noneOf(Capability.class));

    private final EnumSet<Capability> caps;

    private CapabilitySet(EnumSet<Capability> caps) { this.caps = caps; }

    @Override public CapabilitySet toCapabilitySet() { return this; }

    public static CapabilitySet fromNames(Collection<String> names) {
        EnumSet<Capability> caps = EnumSet.noneOf(Capability.class);
        for (String name : names) {
            var predefinedSet = PREDEFINED.get(name);
            var capability = Capability.fromName(name).orElse(null);
            if (capability != null) caps.add(capability);
            else if (predefinedSet != null) caps.addAll(predefinedSet.caps);
            else log.warning("Cannot find capability or capability set with name '%s'".formatted(name));
        }
        return new CapabilitySet(caps);
    }

    public static CapabilitySet unionOf(Collection<CapabilitySet> capSets) {
        EnumSet<Capability> union = EnumSet.noneOf(Capability.class);
        capSets.forEach(cs -> union.addAll(cs.caps));
        return new CapabilitySet(union);
    }

    public static CapabilitySet of(ToCapabilitySet... capabilities) {
        return CapabilitySet.unionOf(Arrays.stream(capabilities).map(ToCapabilitySet::toCapabilitySet).toList());
    }

    public static CapabilitySet of(EnumSet<Capability> caps) { return new CapabilitySet(EnumSet.copyOf(caps)); }
    public static CapabilitySet of(Collection<Capability> caps) { return new CapabilitySet(EnumSet.copyOf(caps)); }
    public static CapabilitySet of(Capability... caps) { return new CapabilitySet(EnumSet.copyOf(List.of(caps))); }
    public static CapabilitySet all() { return ALL_CAPABILITIES; }
    public static CapabilitySet none() { return NO_CAPABILITIES; }

    public boolean hasAll() { return this.caps.equals(ALL_CAPABILITIES.caps); }
    public boolean hasNone() { return this.caps.equals(NO_CAPABILITIES.caps); }
    public boolean has(CapabilitySet caps) { return this.caps.containsAll(caps.caps); }
    public boolean has(Collection<Capability> caps) { return this.caps.containsAll(caps); }
    public boolean has(Capability... caps) {  return this.caps.containsAll(List.of(caps)); }

    public SortedSet<String> toNames() {
        return caps.stream().map(Capability::asString).collect(Collectors.toCollection(TreeSet::new));
    }

    public Set<Capability> asSet() { return Collections.unmodifiableSet(caps); }

    @Override
    public String toString() {
        return "CapabilitySet{" +
                "caps=" + caps +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        CapabilitySet that = (CapabilitySet) o;
        return Objects.equals(caps, that.caps);
    }

    @Override
    public int hashCode() {
        return Objects.hash(caps);
    }
}
